ouroboros_macro/src/covariance_detection.rs

```diff
@@ -40,88 +40,99 @@ pub fn apparent_std_container_type(raw_type: &Type) -> Option<(&'static str, &Ty
             }
             for segment in path.path.segments.iter() {
                 let args = &segment.arguments;
                 if let syn::PathArguments::AngleBracketed(args) = &args {
                     for arg in args.args.iter() {
                         if let syn::GenericArgument::Type(ty) = arg {
-                            if !type_is_covariant(ty, !is_covariant) {
-                                return false;
+                            if all_parameters_are_covariant {
+                                if !type_is_covariant_over_this_lifetime(ty)? {
+                                    return Some(false);
+                                }
+                            } else {
+                                if uses_this_lifetime(ty.to_token_stream()) {
+                                    return None;
+                                }
                             }
                         } else if let syn::GenericArgument::Lifetime(lt) = arg {
-                            if lt.ident.to_string() == "this" && !is_covariant {
-                                return false;
+                            if lt.ident.to_string() == "this" && !all_parameters_are_covariant {
+                                return None;
                             }
                         }
                     }
                 } else if let syn::PathArguments::Parenthesized(args) = &args {
                     for arg in args.inputs.iter() {
-                        if !type_is_covariant(arg, true) {
-                            return false;
+                        if uses_this_lifetime(arg.to_token_stream()) {
+                            return None;
                         }
                     }
                     if let syn::ReturnType::Type(_, ty) = &args.output {
-                        if !type_is_covariant(ty, true) {
-                            return false;
+                        if uses_this_lifetime(ty.to_token_stream()) {
+                            return None;
                         }
                     }
                 }
             }
-            true
+            Some(true)
         }
-        Ptr(ptr) => type_is_covariant(&ptr.elem, in_template),
+        Ptr(ptr) => type_is_covariant_over_this_lifetime(&ptr.elem),
         // Ignore the actual lifetime of the reference because Rust can automatically convert those.
         Reference(rf) => {
-            !in_template && rf.mutability.is_none() && type_is_covariant(&rf.elem, in_template)
+            if rf.mutability.is_some() {
+                Some(!uses_this_lifetime(rf.elem.clone().into_token_stream()))
+            } else {
+                type_is_covariant_over_this_lifetime(&rf.elem)
+            }
         }
-        Slice(sl) => type_is_covariant(&sl.elem, in_template),
-        TraitObject(..) => false,
+        Slice(sl) => type_is_covariant_over_this_lifetime(&sl.elem),
+        TraitObject(..) => None,
```

Comment:
Actually, trait object types containing `'this` should generally be not covariant, so `Some(false)` seems better.

Reply:
Wait, no, one would have to differentiate lifetimes bounds with lifetimes appearing in generic arguments. The former would be _covariant_. I.e.

* covariant `dyn Foo + 'this`
* invariant `dyn Foo<'this>`
* infaviant `dyn Foo<&'this ()>`

Reply:
You don’t have to handle this in this PR.